pyuvsim/source.py

```diff
@@ -149,26 +275,13 @@ def coherency_calc(self, telescope_location):
 
         if NstokesI < self.Ncomponents:
             # If there are any polarized sources, do rotation.
+            rotation_matrix = self._calc_coherency_rotation(telescope_location)
 
-            # First need to calculate the sin & cos of the parallactic angle
-            # See Meeus's astronomical algorithms eq 14.1
-            # also see Astroplan.observer.parallactic_angle method
-
-            polarized_sources = np.where(~Ionly_mask)[0]
-            sinX = np.sin(self.hour_angle)
-            cosX = (np.tan(telescope_location.lat) * np.cos(self.dec)
-                    - np.sin(self.dec) * np.cos(self.hour_angle))
-
-            # shape (2, 2, Ncomponents)
-            rotation_matrix = np.array([[cosX, sinX], [-sinX, cosX]]).astype(float)
-            rotation_matrix = rotation_matrix[..., polarized_sources]
-
-            rotation_matrix_T = np.swapaxes(rotation_matrix, 0, 1)
-            coherency_local[:, :, polarized_sources] = np.einsum(
-                'abx,bcx,cdx->adx', rotation_matrix_T,
-                self.coherency_radec[:, :, polarized_sources],
-                rotation_matrix
-            )
+            for src_ind in range(self.Ncomponents):
```

Comment:
If I'm not mistaken, this part is already vectorized on the master branch and the same syntax can be used here:

```
            polarized_sources = np.where(~Ionly_mask)[0]
            rotation_matrix_T = np.swapaxes(rotation_matrix, 0, 1)
            coherency_local[:, :, polarized_sources] = np.einsum(
                'abx,bcx,cdx->adx', rotation_matrix_T,
                self.coherency_radec[:, :, polarized_sources],
                rotation_matrix
            )
```

Reply:
I don't think that part of the code is run in the master branch. Trying to make that change errors for me.

Reply:
Ignore the above, I just hadn't figured out ellipses in `einsum`. The real problem is that the `rotation_matrix` is different per source. I plan to vectorize it in a future PR.

Reply:
It should already be a separate rotation matrix per source (the "x" axis in the summation string). The problem may be that you have Ncomponents rotation matrices when the einsum only expects them for the polarized sources.
So you'd replaced `rotation_matrix` with `rotation_matrix[:,:,polarized_sources]` and likewise for `.rotation_matrix_T`.

Reply:
The Ncomponents is ok because I'm currently calculating the rotation matrix for all the sources, not just the polarized ones. The real problem was that the transposition puts the source axis first.